modules/gapi/src/streaming/onevpl/engine/preproc/preproc_engine.cpp

```diff
@@ -0,0 +1,491 @@
+        }
+    } catch (const std::exception&) {
+        MFXClose(mfx_vpp_session);
+        GAPI_Assert(false && "Cannot init preproc resources");
+    }
+
+    // create engine session after all
+    session_ptr_type sess_ptr = register_session<session_type>(mfx_vpp_session,
+                                                               mfxVPPParams);
+    sess_ptr->init_surface_pool(vpp_out_pool_key);
+    sess_ptr->swap_surface(*this);
+
+    bool inserted = preproc_session_map.emplace(mfxVPPParams.vpp.In, sess_ptr).second;
+    GAPI_Assert(inserted && "preproc session is exist");
+    GAPI_LOG_INFO(nullptr, "VPPPreprocSession created, total sessions: " << preproc_session_map.size());
+    return pp_session::create(std::static_pointer_cast<EngineSession>(sess_ptr));
+}
+
+void VPPPreprocEngine::on_frame_ready(session_type& sess,
+                                      mfxFrameSurface1* ready_surface)
+{
+    GAPI_LOG_DEBUG(nullptr, "[" << sess.session << "], frame ready");
+
+    // manage memory ownership rely on acceleration policy
+    ready_surface->Data.Locked--;  // TODO -S- workaround
+    VPLAccelerationPolicy::FrameConstructorArgs args{ready_surface, sess.session};
+    auto frame_adapter = acceleration_policy->create_frame_adapter(sess.vpp_pool_id,
+                                                                   args);
+    ready_frames.emplace(cv::MediaFrame(std::move(frame_adapter)), sess.generate_frame_meta());
+
+    // pop away synced out object
+    sess.vpp_out_queue.pop();
+}
+
+VPPPreprocEngine::session_ptr
+VPPPreprocEngine::initialize_session(mfxSession,
+                                     const std::vector<CfgParam>&,
+                                     std::shared_ptr<IDataProvider>) {
+    return {};
+}
+
+cv::MediaFrame VPPPreprocEngine::run_sync(const pp_session& sess, const cv::MediaFrame& in_frame) {
+
+    std::shared_ptr<EngineSession> pp_sess_impl = sess.get<EngineSession>();
+    if (!pp_sess_impl) {
+        // bypass case
+        return in_frame;
+    }
+    session_ptr_type s = std::static_pointer_cast<session_type>(pp_sess_impl);
+    GAPI_DbgAssert(s && "Session is nullptr");
+    GAPI_DbgAssert(is_applicable(in_frame) &&
+                   "VPP preproc is not applicable for the given frame");
+    BaseFrameAdapter *vpl_adapter = in_frame.get<BaseFrameAdapter>();
+    if (!vpl_adapter) {
+        GAPI_LOG_WARNING(nullptr, "VPP preproc is inapplicable for a given frame. "
+                                  "Make sure the frame is collected using onevpl::GSource");
+        throw std::runtime_error("VPP preproc is inapplicable for given frame");
+    }
+
+    // schedule decoded surface into preproc queue
+    session_type::op_handle_t in_preproc_request {nullptr,
+                                                  vpl_adapter->get_surface()->get_handle()};
+    s->sync_in_queue.emplace(in_preproc_request);
+    remember_decode_frame(vpl_adapter->get_surface()->get_handle(), in_frame);
```

Comment:
i don't think about that..
But as far i remember `pipeline`-lambdas was designed as pure- and is cutting from the other world and cannot receive different kind of data across `session`

actually it can be passed as another aggregated member of `session_type::op_handle_t`  as well.
Moreover a subsequent PR  about ROI extention follows this approach: https://github.com/opencv/opencv/pull/21554/files#diff-d281efabce2ff069582bd99d9c40e910c4386b65d60010ed9e4b106f4f2b237bR384

@rgarnov  Let's move Frame into op_handle_t, what do you think ?